### ISSUE TITLE
[RFC] Changing the set of ciphers

### ISSUE THREAD

**Issue by beta-ziliani** [crystal-lang/crystal#13727]
This is an extraction of #13686.

Currently Crystal fixes the set of ciphers based on Mozilla's recommendation ([currently 5.7 of the old, most compatible, configuration](https://ssl-config.mozilla.org/)).

However, the recommended behavior by RedHat and Fedora is to leave it to the system ([docs.fedoraproject.org/en-US/packaging-guidelines/CryptoPolicies](https://docs.fedoraproject.org/en-US/packaging-guidelines/CryptoPolicies/)). This makes sense: an up-to-date system will be better prepared than an app that was compiled at an older time.

 This said, an app compiled with a recent version of Crystal will be safer running on an out-of-date OS, so there's a trade-off between how much we trust a client's computer, and how much we trust Crystal and Crystal apps to be up-to-date.
 
 An alternative could be to query the list of ciphers from the OS and remove those that are known to be problematic (although I don't know how to easily get this information in order to get it updated).
 
 Another alternative is to add an environment variable to set the list (with `""` or "SYSTEM" meaning not setting the list).

**Comment by straight-shoota** [comment]
I suppose using the default configuration may be the best solution. It's simple and unsurprising.

I don't see much value in removing problematic ciphers from that default configuration. This only really works as a snap shot of whatever is considered problematic in that specific version of Crystal. So it's incalculable and does not provide any forward security. So I'd rather avoid this additional complexity.
Also when you'rerunning an old version of OpenSSL, individual ciphers are likely not the only problem.

Having more out-of-the-box options to specify the configuration at runtime could be a helpful addition.

**Comment by straight-shoota** [comment]
It might be relevant to think about client and server contexts separately, though. It's not clear if the Fedora documentation explicitly applies to both.

I think there are good reasons to treat them differently and the [Mozilla recommendations](https://wiki.mozilla.org/Security/Server_Side_TLS) are specifically for server applications.

Currently, we're using Mozilla's `OLD` configuration for client contexts (even if it's meant for servers), and `INTERMEDIATE` for server contexts.

**Comment by ysbaddaden** [comment]
We may want to enhance the integration with [SSL_CTX_set_min_proto_version](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_min_proto_version.html) (available since OpenSSL 1.1.0) and keep it unspecified by default.

That would allow to set TLSv1 or TLSv1.1 when backward compatibility is needed and only enable TLSv1.3 for "modern" standards.

**Comment by ysbaddaden** [comment]
I couldn't find a nice tool to test client SSL connections (despite being just as important). Here's what I found using an insecure context (Ubuntu 20.04 with OpenSSL 1.1.1f):

- I can connect to TLSv1.2 and TLSv1.3 servers, while I can't connect to TLSv1 or TLSv1.1 servers :+1: ;
- The exposed ciphers list is conservative (roughly like the "old" mozilla ssl-config) :shrug:

The default context still seems a bit more secure, as it disables session renegotiation, disables everything below TLSv1.2, and more.

Again it would be interesting to see what happens on other systems, but it sounds like OpenSSL finally got some acceptable defaults starting from v1.1 at least :eyes: 

```crystal
require "http/client"
tls = OpenSSL::SSL::Context::Client.insecure
tls.set_default_verify_paths
HTTP::Client.get(HOST, PORT, tls: tls).get(PATH)
```
